feat(useStorageValue): add custom parse and stringify functions to options

dantman opened the pull request:
## Overview

`parse` and `stringify` options have been added to useStorageValue to allow a custom parse and stringify to be used in place of the default JSON based parse/stringify.

Sometimes a custom format or a custom replacer function are needed for storing values in localstorage.

## Checklist

- [x] Have you read [contribution guideline](../../CONTRIBUTING.md)?
- [x] Does the code have comments in hard-to-understand areas?
- [x] Is there an existing issue for this PR?
  - #942
- [x] Have the files been linted and formatted?
- [x] Have the docs been updated?
- [x] Have the tests been added to cover the changes?
- [x] Have you run the tests locally to confirm they pass?

ArttuOll commented:
Thanks for the PR!

Could you please reimplement your changes on top of the new `useStorageValue` implementation? I tried to resolve the conflicts this branch has with master, but noticed that they are more complicated than what I currently have time for.

After that we will be able to do a review.

xobotyi commented:
I have concerns about this change since you'll be able to configure different hook instances pointing to same key having different serializers.

ArttuOll commented:
> I have concerns about this change since you'll be able to configure different hook instances pointing to same key having different serializers.

Maybe we could address this by adding a warning about this to the documentation? It would be the developer's responsibility to make sure that the situation you described does not happen.

xobotyi commented:
presumably you're right

dantman commented:
> Could you please reimplement your changes on top of the new `useStorageValue` implementation? I tried to resolve the conflicts this branch has with master, but noticed that they are more complicated than what I currently have time for.

Ok, I'll look into that next week.

dantman commented:
I've updated the code to handle the newly introduced types to the serializes and updated the documentation with a warning and some typo fixes.

I did have to remove the type from DEFAULT_OPTIONS since it was injecting some extra `null` types into the serializers that was breaking the options version of them.

ArttuOll commented on src/useSessionStorageValue/__docs__/story.mdx:
```diff
@@ -59,6 +62,8 @@ function useSessionStorageValue<
   - **initializeWithValue** _`boolean`_ _(default: true)_ - Fetch storage value on first render. If
     set to `false` will make the hook yield `undefined` on first render and defer fetching of the
     value until effects are executed.
+  - **parse** _`function`_ - Custom parse function to use instead of JSON.parse
+  - **stringify** _`function`_ - Custom stringify function to use instead of JSON.stringify
```
Better to include the type signatures of the functions here instead of just saying `function`.

ArttuOll reviewed (changes requested):
Everything works, just one change to the docs and maybe an improvement to the types and we're good to go!

dantman commented on src/useStorageValue/useStorageValue.ts:
```diff
@@ -135,6 +145,14 @@ export function useStorageValue<
   options?: UseStorageValueOptions<Type, Initialize>
 ): UseStorageValueResult<Type, Default, Initialize> {
   const optionsRef = useSyncedRef({ ...DEFAULT_OPTIONS, ...options });
+  const parse = (str: string | null, fallback: Type | null): Type | null => {
+    const parseFunction = optionsRef.current.parse ?? defaultParse;
+    return parseFunction(str, fallback);
+  };
+  const stringify = (data: unknown): string | null => {
+    const stringifyFunction = optionsRef.current.stringify ?? defaultStringify;
+    return stringifyFunction(data);
+  };
```
The only way to do so that I could see, would be to use `Pick` to explicitly target the options that are used. But that ends up being pointlessly verbose, needing to update the list of keys each time you add a default. I could revisit typing useSyncedRef so DEFAULT_OPTIONS is indirectly checked. I tried that before and had issues, but maybe that was before I switched to these parse/stringify wrappers to deal with other type issues.

ArttuOll replied:
I think we're good to go like this.

ArttuOll reviewed (approved):
I merged master and added the missing function types to the documentation.

@xobotyi This looks good to me now. Am I allowed to merge PR's I've reviewed or do you want to also review everything?

xobotyi commented:
LGTM

xobotyi commented:
:tada: This PR is included in version 19.1.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@react-hookz/web/v/19.1.0)
- [GitHub release](https://github.com/react-hookz/web/releases/tag/v19.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
